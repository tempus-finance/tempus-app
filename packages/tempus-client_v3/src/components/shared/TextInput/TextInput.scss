$textinput-border-radius: 4px;

.tc__text-input {
  .tc__text-input__container {
    position: relative;
    display: flex;
    justify-content: center;
    background: var(--textInputBackground);
    border: 1px solid var(--textInputBorder);
    border-radius: $textinput-border-radius;
    box-sizing: border-box;
    box-shadow: var(--downWeakShadow);
    height: 40px;

    input {
      background: transparent;
      border: none;
      border-radius: $textinput-border-radius;
      padding: 0 12px;
      height: 38px;

      &:focus-visible {
        outline: none;
      }
    }
  }

  &.tc__text-input__focused {
    .tc__text-input__container {
      &::after {
        content: ' ';
        position: absolute;
        width: 100%;
        height: 2px;
        top: 38px;
        background: var(--textCaption);
      }
    }
  }

  &.tc__text-input__error {
    .tc__text-input__container {
      &::after {
        content: ' ';
        position: absolute;
        width: 100%;
        height: 2px;
        top: 38px;
        background: var(--textError);
      }
    }
  }

  &.tc__text-input__disabled {
    .tc__text-input__container {
      background: var(--textInputBackgroundDisabled);
      border: 1px solid var(--textInputBorderDisabled);
    }
  }
}
