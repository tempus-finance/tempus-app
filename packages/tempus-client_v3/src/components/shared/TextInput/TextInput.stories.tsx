import { ComponentStory } from '@storybook/react';
import React, { FC, useCallback, useState } from 'react';
import TextInput, { TextInputProps } from './TextInput';
import Logo from '../Logo';

export default {
  title: 'TextInput',
  component: TextInput,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Wrapper: FC<TextInputProps> = props => {
  const [value, setValue] = useState<string>(props.value ?? '');
  const onChange = useCallback((val: string) => setValue(val), []);
  return <TextInput {...props} value={value} onChange={onChange} />;
};

const Template: ComponentStory<typeof TextInput> = args => (
  <div style={style}>
    <Wrapper {...args} />
  </div>
);

export const TextInputRaw = Template.bind({});
TextInputRaw.args = {
  placeholder: 'placeholder',
  pattern: '',
  disabled: false,
  debounce: false,
  startAdornment: null,
  endAdornment: null,
};

export const TextInputNormal = Template.bind({});
TextInputNormal.args = {
  label: 'text input',
  placeholder: 'placeholder',
  pattern: '',
  disabled: false,
  debounce: false,
  startAdornment: null,
  endAdornment: null,
};

export const TextInputCaption = Template.bind({});
TextInputCaption.args = {
  label: 'text input',
  placeholder: 'placeholder',
  pattern: '',
  caption: 'caption here',
  disabled: false,
  debounce: false,
  startAdornment: null,
  endAdornment: null,
};

export const TextInputError = Template.bind({});
TextInputError.args = {
  label: 'text input',
  placeholder: 'placeholder',
  pattern: '',
  caption: 'caption here',
  error: 'error here',
  disabled: false,
  debounce: false,
  startAdornment: null,
  endAdornment: null,
};

export const TextInputDisabled = Template.bind({});
TextInputDisabled.args = {
  label: 'text input',
  placeholder: 'placeholder',
  pattern: '',
  caption: 'caption here',
  disabled: true,
  debounce: false,
  startAdornment: null,
  endAdornment: null,
};

export const TextInputWithStartAdornment = Template.bind({});
TextInputWithStartAdornment.args = {
  label: 'text input',
  placeholder: 'placeholder',
  pattern: '',
  caption: 'caption here',
  disabled: false,
  debounce: false,
  startAdornment: <Logo type="token-ETH" />,
  endAdornment: null,
};

export const TextInputWithEndAdornment = Template.bind({});
TextInputWithEndAdornment.args = {
  label: 'text input',
  placeholder: 'placeholder',
  pattern: '',
  caption: 'caption here',
  disabled: false,
  debounce: false,
  startAdornment: null,
  endAdornment: <Logo type="token-ETH" />,
};
