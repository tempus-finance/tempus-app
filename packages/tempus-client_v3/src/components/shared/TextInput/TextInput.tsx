import React, { ReactNode, FocusEvent, FC, useCallback, useMemo, useState, memo } from 'react';
import BaseInput from '../BaseInput';
import Typography from '../Typography';
import '../Shadow';

import './TextInput.scss';

export interface TextInputProps {
  label?: string;
  value?: string;
  placeholder?: string;
  pattern?: string;
  caption?: string;
  error?: string;
  disabled?: boolean;
  debounce?: boolean | number;
  startAdornment?: ReactNode;
  endAdornment?: ReactNode;
  onChange?: (value: string) => void;
}

let idCounter = 0;

const TextInput: FC<TextInputProps> = props => {
  const {
    label,
    value,
    placeholder,
    pattern,
    caption,
    error,
    disabled,
    debounce = false,
    startAdornment,
    endAdornment,
    onChange,
  } = props;
  const [focused, setFocused] = useState<boolean>(false);
  const id = useMemo(() => `text-input-${idCounter++}`, []);

  const handleFocus = useCallback(() => setFocused(true), []);
  const handleBlur = useCallback(() => setFocused(false), []);

  return (
    <div
      className={`tc__text-input ${
        disabled
          ? 'tc__text-input__disabled'
          : error
          ? 'tc__text-input__error'
          : focused
          ? 'tc__text-input__focused'
          : ''
      }`}
    >
      {label && (
        <label htmlFor={id}>
          <Typography variant="body-secondary" weight="bold">
            {label}
          </Typography>
        </label>
      )}
      <div className="tc__text-input__container">
        {startAdornment}
        <Typography variant="body-primary" weight="regular" color={disabled ? 'text-disabled' : undefined}>
          <BaseInput
            id={id}
            value={value}
            placeholder={placeholder}
            pattern={pattern}
            debounce={debounce}
            disabled={disabled}
            onChange={onChange}
            onFocus={handleFocus}
            onBlur={handleBlur}
          />
        </Typography>
        {endAdornment}
      </div>
      {caption && !error && (
        <Typography variant="body-secondary" weight="medium" color={focused ? 'text-caption' : undefined}>
          {caption}
        </Typography>
      )}
      {error && (
        <Typography variant="body-secondary" weight="medium" color="text-error">
          {error}
        </Typography>
      )}
    </div>
  );
};

export default memo(TextInput);
