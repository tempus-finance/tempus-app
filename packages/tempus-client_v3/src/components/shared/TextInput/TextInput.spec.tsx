import { fireEvent, render } from '@testing-library/react';
import { FC, useState } from 'react';
import { act } from 'react-dom/test-utils';
import TextInput, { TextInputProps } from './TextInput';

const mockOnChange = jest.fn();
const mockOnDebounceChange = jest.fn();

const defaultProps: TextInputProps = {
  label: 'text input label',
  value: '',
  placeholder: 'text input placeholder',
  pattern: '',
  caption: '',
  error: '',
  disabled: false,
  startAdornment: null,
  endAdornment: null,
  onChange: mockOnChange,
  onDebounceChange: mockOnDebounceChange,
};

const Wrapper: FC<TextInputProps> = props => {
  const [value, setValue] = useState<string>(props.value ?? '');
  mockOnChange.mockImplementation((val: string) => setValue(val));
  return <TextInput {...props} value={value} />;
};

const subject = (props: TextInputProps) => render(<Wrapper {...props} />);

describe('TextInput', () => {
  it('renders a text input with label', () => {
    const { container, getByRole, queryByLabelText } = subject(defaultProps);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with value', () => {
    const props = { ...defaultProps, value: 'value' };
    const { container, getByRole, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(input).toHaveValue(props.value);

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with pattern', () => {
    const props = { ...defaultProps, pattern: '[0-9,]*[.]?[0-9]*' };
    const { container, getByRole, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(input.getAttribute('pattern')).toBe(props.pattern);

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with caption', () => {
    const props = { ...defaultProps, caption: 'this is a caption' };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);
    const caption = queryByText(props.caption);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(caption).not.toBeNull();

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with caption and error', () => {
    const props = { ...defaultProps, caption: 'this is a caption', error: 'this is an error' };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);
    const caption = queryByText(props.caption);
    const error = queryByText(props.error);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(caption).toBeNull();
    expect(error).not.toBeNull();

    expect(textInput).toMatchSnapshot();
  });

  it('renders a disabled text input', () => {
    const props = { ...defaultProps, disabled: true };
    const { container, getByRole, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(input.getAttribute('disabled')).not.toBeNull();

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with startAdornment', () => {
    const props = { ...defaultProps, startAdornment: <div>startAdornment</div> };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);
    const startAdornment = queryByText('startAdornment');

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(startAdornment).not.toBeNull();

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with endAdornment', () => {
    const props = { ...defaultProps, endAdornment: <div>endAdornment</div> };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);
    const endAdornment = queryByText('endAdornment');

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(endAdornment).not.toBeNull();

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with focus', () => {
    const { container, getByRole, queryByLabelText } = subject(defaultProps);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.focus(input);

    expect(textInput).toMatchSnapshot();
  });

  it('renders a text input with focus with caption', () => {
    const props = { ...defaultProps, caption: 'this is a caption' };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const textInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);
    const caption = queryByText(props.caption);

    expect(textInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(label).not.toBeNull();
    expect(caption).not.toBeNull();

    fireEvent.focus(input);

    expect(textInput).toMatchSnapshot();
  });

  it('type text and then blur in text input with debounce will trigger onChange and onDebounceChange immediately', () => {
    jest.useFakeTimers();
    const props = { ...defaultProps, value: undefined, debounce: true };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });
    expect(mockOnChange).toHaveBeenCalledWith('abcd');
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('abcd');
    act(() => {
      jest.advanceTimersByTime(100);
    });
    fireEvent.blur(input);
    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');

    jest.useRealTimers();
  });
});
