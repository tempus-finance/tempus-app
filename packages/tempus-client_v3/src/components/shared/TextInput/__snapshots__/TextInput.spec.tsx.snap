// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders a disabled text input 1`] = `
<div
  class="tc__text-input tc__text-input__disabled"
>
  <label
    for="text-input-6"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        disabled=""
        id="text-input-6"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  
  
</div>
`;

exports[`TextInput renders a text input with caption 1`] = `
<div
  class="tc__text-input "
>
  <label
    for="text-input-4"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-4"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
  >
    this is a caption
  </div>
  
</div>
`;

exports[`TextInput renders a text input with caption and error 1`] = `
<div
  class="tc__text-input tc__text-input__error"
>
  <label
    for="text-input-5"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-5"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(255, 15, 15); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
  >
    this is an error
  </div>
</div>
`;

exports[`TextInput renders a text input with endAdornment 1`] = `
<div
  class="tc__text-input "
>
  <label
    for="text-input-8"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-8"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
    <div>
      endAdornment
    </div>
  </div>
  
  
</div>
`;

exports[`TextInput renders a text input with focus 1`] = `
<div
  class="tc__text-input tc__text-input__focused"
>
  <label
    for="text-input-9"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-9"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  
  
</div>
`;

exports[`TextInput renders a text input with focus with caption 1`] = `
<div
  class="tc__text-input tc__text-input__focused"
>
  <label
    for="text-input-10"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-10"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(245, 172, 55); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
  >
    this is a caption
  </div>
  
</div>
`;

exports[`TextInput renders a text input with label 1`] = `
<div
  class="tc__text-input "
>
  <label
    for="text-input-1"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-1"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  
  
</div>
`;

exports[`TextInput renders a text input with pattern 1`] = `
<div
  class="tc__text-input "
>
  <label
    for="text-input-3"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-3"
        pattern="[0-9,]*[.]?[0-9]*"
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  
  
</div>
`;

exports[`TextInput renders a text input with startAdornment 1`] = `
<div
  class="tc__text-input "
>
  <label
    for="text-input-7"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div>
      startAdornment
    </div>
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-7"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
  
  
</div>
`;

exports[`TextInput renders a text input with value 1`] = `
<div
  class="tc__text-input "
>
  <label
    for="text-input-2"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      text input label
    </div>
  </label>
  <div
    class="tc__text-input__container"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      <input
        class="tc__base-input"
        id="text-input-2"
        pattern=""
        placeholder="text input placeholder"
        type="text"
        value="value"
      />
    </div>
  </div>
  
  
</div>
`;
